modules/cef/format-cef-extension.c

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright (c) 2015 Balabit
+ * Copyright (c) 2015 bkil.hu
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ */
+
+#include "filter/filter-expr.h"
+#include "filter/filter-expr-parser.h"
+#include "cfg.h"
+#include "value-pairs.h"
+#include "vptransform.h"
+#include "syslog-ng.h"
+#include "utf8utils.h"
+#include "format-cef-extension.h"
+
+typedef struct _TFCefState
+{
+  TFSimpleFuncState super;
+  ValuePairs *vp;
+} TFCefState;
+
+static gboolean
+tf_cef_prepare(LogTemplateFunction *self, gpointer s, LogTemplate *parent,
+		gint argc, gchar *argv[],
+		GError **error)
+{
+  TFCefState *state = (TFCefState *)s;
+
+  state->vp = value_pairs_new_from_cmdline (parent->cfg, argc, argv, error);
+  if (!state->vp)
+    return FALSE;
+
+  return TRUE;
+}
+
+typedef struct
+{
+  gboolean need_separator;
+  GString *buffer;
+  const LogTemplateOptions *template_options;
+} cef_state_t;
+
+static inline void
+g_string_append_escaped (GString *escaped_string, const char *str)
+{
+  const gchar *char_ptr = str;
+
+  while (*char_ptr)
+    {
+      gunichar uchar = g_utf8_get_char_validated (char_ptr, -1);
+
+      switch (uchar)
+        {
+        case '=':
+          g_string_append (escaped_string, "\\=");
+          break;
+        case '\n':
+          g_string_append (escaped_string, "\\n");
+          break;
+        case '\r':
+          g_string_append (escaped_string, "\\r");
+          break;
+        case '\\':
+          g_string_append (escaped_string, "\\\\");
+          break;
+        default:
+          g_string_append_unichar (escaped_string, uchar);
+          break;
+        }
+      char_ptr = g_utf8_next_char (char_ptr);
+    }
+}
+
+static gboolean
+tf_cef_append_value(const gchar *name, const gchar *value,
+                     cef_state_t *state)
+{
+  if (state->need_separator)
+    g_string_append_c(state->buffer, ' ');
+
+  g_string_append_escaped(state->buffer, name);
```

Comment:
The name part has some useful restrictions:
* names are single words,
* contain only alphanumeric characters.

That means, we don't need to escape them as the values. I'd use them as they are, maybe log something into the internal messages if the current character is invalid. Or we can drop these characters, turning the invalid name into a valid one (But what if the name contains only these characters? ...).